File: src/main/java/com/github/takemikami/intellij/plugin/externallinters/TextOffsetDetector.java
package com.github.takemikami.intellij.plugin.externallinters;

import java.io.BufferedReader;
import java.io.StringReader;
import java.util.Arrays;
import org.apache.commons.lang3.ArrayUtils;

/**
 * Text offset detector, from line-number and column-number to offset.
 */
public class TextOffsetDetector {

  private final Integer[] linesLength;
  private final int maxLength;

  /**
   * TextOffsetDetector constructor.
   *
   * @param text text body
   */
  public TextOffsetDetector(String text) {
    maxLength = text.length();
    BufferedReader buff = new BufferedReader(new StringReader(text));
    linesLength = buff.lines().map(String::length).toArray(Integer[]::new);
  }

  /**
   * Return offset in text body.
   *
   * @param line line number
   * @param col  column number
   * @return offset
   */
  public int getOffset(int line, int col) {
    int offset = Arrays.stream(
        ArrayUtils.subarray(linesLength, 0, line - 1)
    ).map(l -> (int) l + 1).reduce(Integer::sum).orElse(0) + col - 1;
    return Math.min(offset, maxLength - 1);
  }

  public int getOffset(int line) {
    return getOffset(line, 1);
  }

}

File: src/main/kotlin/com/github/takemikami/intellij/plugin/externallinters/TextOffsetDetector.kt
package com.github.takemikami.intellij.plugin.externallinters

import java.io.BufferedReader
import java.io.StringReader
import kotlin.math.min

/**
 * Text offset detector, from line-number and column-number to offset.
 */
class TextOffsetDetector(text: String) {
    private val maxLength = text.length
    val buff = BufferedReader(StringReader(text))
    private val linesLength = buff.lines().map { obj: String? -> obj!!.length }.toList() as List<Int>

    /**
     * Return offset in text body.
     *
     * @param line line number
     * @param col  column number
     * @return offset
     */
    fun getOffset(
        line: Int,
        col: Int,
    ): Int {
        val offsetLine =
            linesLength
                .take(line - 1)
                .map { l -> l + 1 }
                .reduceOrNull { a, b -> a + b }
                ?: 0
        val offset = offsetLine + col - 1
        return min(offset, maxLength - 1)
    }

    fun getOffset(line: Int): Int {
        return getOffset(line, 1)
    }
}
